prime-router/docs/proposals/lineage/README.md

```diff
@@ -0,0 +1,123 @@
+
+## Proposed Design
+
+Please [pull up the ER diagram](./lineage-er-diagram.jpeg) for the detailed proposed schema.
+
+(Note: [Original lucid chart diagram is here](https://lucid.app/lucidchart/7b81cccb-91a0-44b3-b17f-90b9b0c8304e/edit?beaconFlowId=332110D5160D6847&page=0_0#))
+
+### Notes on the Proposed Schema
+
+- The separation of Task from Report allows us to have Tasks that don't necessarily create Reports.
+- Addition of the `TASK.task_result` fills a gap where we're not tracking the error messages sent back to users.
+- Apparently I can't name the REPORT table 'REPORT', due to a limitation in Flyway/jooq.  Hence the weird name `REPORT_FILE`.   I also thought about `REPORT_IN`, `REPORT_DATA`, `REPORT_REC`.  Haven't sunk as low as `REPURT`, but I'm on my way there...
+- Many values in REPORT_FILE will be empty depending on the task creating the report.  For example, all the sending_* fields are only filled in for an initially received report.   All the receiving_* fields are only filled in for reports sent externally.
+- might need a report_status in the future.  For example, `wiped` or `archived`, to explain why a blob might have disappeared.
+- The primary key to use for ITEM is tricky.  The sender gives us the `tracking_element`, but we have little control over how robust that is.  Because of this, all queries on `tracking_element` must also filter on report_id.
+- So there's an open question:  should we require global uniqueness of `tracking_elements`?   It would be wonderfully convenient to use `tracking_elements` as primary keys in our database.    However, they would have to be globally unique, because report data can get merged together in arbitrary ways in steps downstream from Receive.
+- There's subtlety with Report parent/child relationships.   Because any given task might take in a bunch of reports, and produce a bunch of reports, there's no guarantee that any of the data in output Report Y came from input Report X.  However we're tracking it in the REPORT_LINEAGE table as if every input report fed every output report.  In practice, its very hard to tell if that's true, in general.
+
+### How this design helps build trust
+
+- Fine-grained lineage tracking at the ITEM level, with ability to query both "backward" and "forward"
+- Improved tracking of sender and receiver ownership
+- Tracking versions of the functions that created the data, and SHA hash of every report.
+- Next steps: Have an easy-to-use API that delivers history
+- Next steps: Have an easy-to-use API that can retrieve the data in an individual ITEM. (?)
+
+### Rollout plan:
+
+1.  Build out the new schema in parallel with the existing TASK, TASK_SOURCE tables.
+2.  Wire the REPORT_FILE and REPORT_LINEAGE tables first, then the ITEM and ITEM_LINEAGE.
+3.  Code into the validation step a requirement for uniqueness of tracking Ids within each Received report. 
+4.  Migrate existing data to the new tables, and retire or modify the old ones.
```

Comment:
A few questions on the rollout plan:

* What do you mean by built in parallel? Will the schema be deployed with no data until we're ready to migrate? Will data exist in both schemas for a period of time?
* When we do a cut over to the new schema, how do we plan out migrating the old data? Will there be a migration process that copies the data over time? A one-off script that runs (with I assume some system downtime)?
* What's the amount of data we're looking to migrate?

Reply:
@ronaldheft-usds Good questions - thanks!   I've modified the README, but here's the quick answers:
- data will exist in both for a period of time.
- we'll do a quick one-off script.
- Tiny amount of data.   less than 1000 rows at the moment.